### ISSUE TITLE
What's the right way to add asset from a plugin

### ISSUE THREAD

**Issue by lgordey** [webpack/webpack#1175]
My webpack plugin created new file e.g. `ccc.svg`.
How can i add it to assets array from apply plugin section so it will appear in output path as:

``` javascript
MyPlugin.prototype.apply = function(compiler) {

   var assetName = 'ccc.svg',
       assetContent = '<svg>123</svg>';

   compiler.plugin('add-asset', function(assetsArray) {
      assetsArray[assetName] = assetContent;
   }
}
```

So if I have structure like:

inputPath
--- assets
------- aaa.svg
-------  bbb.svg

and after webpack compilation:

outputPath
--- assets
------- asd2131wdsf32fwer2.aaa.svg
------- gfdg8dfgdfg7dfg8dfg.bbb.svg
------- kj234hf24fwer9ger8f2.ccc.svg

It should be simple operation, but i couldn't find a solution.

**Comment by lgordey** [comment]
Of course i can do something like:
```javascript
var assetName = 'ccc.svg',
    outputPath = './output/assets',
    assetContent = '<svg>123</svg>';

compiler.plugin('done', function(stats) {

   fs.writeFileSync(
      path.join(outputPath, stats.hash + '.' + assetName),
      assetContent
   );

});
```

But in this case `kj234hf24fwer9ger8f2.ccc.svg` file will not appear in [manifest](https://github.com/nickjj/manifest-revision-webpack-plugin) and hash from
```javascript
module: {
    loaders: [
        { test: /\.svg$/i, loaders: ['file?context=' + rootAssetPath + '&name=[hash].[name].[ext]'] },
    ]
},
```
is not the same as `stats.hash`

**Comment by sokra** [comment]
add it to `compilation.assets`

i. e. 
``` js
compiler.plugin("emit", function(compilation) {
  compilation["my-asset.svg"] = {
    source: function() { return new Buffer(); },
    size: function() { return 5; }
  };
});
```

**Comment by lgordey** [comment]
Thank you, @sokra, but if I understand you correctly - this not working.

I wrote like this:
```javascript
var source = '<svg>123</svg>'

compiler.plugin('emit', function(compilation, callback) {
  compilation.assets['my-sprite.svg'] = {
    source: function() { return new Buffer(source) },
    size: function() { return Buffer.byteLength(source) }
  }

  callback();
});
```
In this way my new asset `my-sprite.svg` added to `compilation.assets` array and saved to output path and that's good!  
But there is no hash and no file `my-sprite.svg` in [manifest.json](https://github.com/nickjj/manifest-revision-webpack-plugin). So in conclusion it's not working yet.
And is it really right way to set the size of asset i'v just added? That's strange

**Comment by sokra** [comment]
@nickjj

**Comment by nickjj** [comment]
@lgordey https://github.com/nickjj/manifest-revision-webpack-plugin

I wrote that plugin for specifically what you're trying to do. 

here's also a presentation on using it with Flask here:
https://www.youtube.com/watch?v=rQoJcqx42pM

At some point during the presentation it goes over the `manifest.json` file it spits out.

**Comment by lgordey** [comment]
@nickjj I'm already using your plugin https://github.com/nickjj/manifest-revision-webpack-plugin

It works well with all assets that were there (in input path) from the start.

But my problem that i have another plugin that should dynamically add a new asset (i wrote an example higher). 

And as I'v already say - that's not working, because what i'm setting in `compilation.assets` is not appearing in manifest. 

And files in `compilation.assets` already have chunkhash, and my dynamically added asset - haven't.

**Comment by lgordey** [comment]
@nickjj @sokra Hi guys! 

I made a repo with issue representation.

https://github.com/lgordey/webpack-plugin

I think it will help to resolve it.

Thank you!

**Comment by pracucci** [comment]
I'm having the same exact issue with Webpack 2. Are you aware of any way to accomplish it with Webpack 2?

**Comment by halfninja** [comment]
@pracucci I don't know if you did the same thing as me, but I didn't realise that the `callback` was mandatory. I just got this working in Webpack 2: https://gist.github.com/halfninja/4dd41e8160078b7e0c3b3d410a2189e0

**Comment by chickenwing** [comment]
Did anyone manage to solve this? Having the same problem right now...

**Comment by Yankovsky** [comment]
@chickenwing check this out https://webpack.js.org/contribute/writing-a-plugin/#example

**Comment by lubomir-chatoryiskyi** [comment]
I've faced the same issue but I use [webpack-assets-manifest](https://github.com/webdeveric/webpack-assets-manifest) to generate manifest.json. 
Don't know is it correct or no - my workaround is to add file generated by plugin to `compilation.assets` and `compilation.chunks[chunkIndex].files`. My generated files were related to original chunks so, I looked for chunk index at first
`const chunkIndex = compilation.chunks.findIndex(chunk => options.from.indexOf(chunk.name) !== -1);`
and after pushed file to array of files
`compilation.chunks[chunkIndex].files.push(filePath);`
After it manifest plugin started to see files generated by plugin and added them to manifest.
